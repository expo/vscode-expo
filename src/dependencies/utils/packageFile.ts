import { findNodeAtLocation, getNodeValue, Node, parse, parseTree, Range } from 'jsonc-parser';

export type PackageFile = { tree: Node; data: any };

/** The package.json cache that keeps track of a single file */
let packageConfigCache: Record<string, PackageFile> = {};

export function getPackageConfig(content: string): PackageFile | undefined {
  if (content in packageConfigCache) {
    return packageConfigCache[content];
  }

  const tree = parseTree(content);
  const data = tree && parse(content);

  if (!tree || !data) {
    return undefined;
  }

  packageConfigCache = {
    [content]: { tree, data },
  };

  return { tree, data };
}

export function findDependencyVersion(
  packageFile: PackageFile,
  dependencyName: string
): string | undefined {
  for (const type of ['dependencies', 'peerDependencies', 'devDependencies']) {
    if (packageFile.data[type]?.[dependencyName]) {
      return packageFile.data[type][dependencyName];
    }
  }
}

export function getPackageDependencies(
  packageFile: PackageFile,
  dependencyType: 'dependencies' | 'devDependencies' | 'peerDependencies'
) {
  const block = findNodeAtLocation(packageFile.tree, [dependencyType]);
  if (!block?.children?.length) {
    return [];
  }

  return block.children.map((node) => {
    if (node.children?.length === 2) {
      return {
        nameNode: node.children[0],
        nameValue: getNodeValue(node.children[0]),
        versionNode: node.children[1],
        versionValue: getNodeValue(node.children[1]),
        versionRange: {
          offset: node.children[1].offset + 1, // Remove starting quote
          length: node.children[1].length - 2, // Remove ending quote
        },
      };
    }
  }) as [
    {
      nameNode: Node;
      nameValue: string;
      versionNode: Node;
      versionValue: string;
      versionRange: Range;
    }
  ];
}
