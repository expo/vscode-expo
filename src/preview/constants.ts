export enum ExpoConfigType {
  PREBUILD = 'prebuild',
  INTROSPECT = 'introspect',
  PUBLIC = 'public',
}

export enum PreviewCommand {
  OpenExpoFilePrebuild = 'expo.config.prebuild.preview',
  OpenExpoFileJsonPrebuild = 'expo.config.prebuild.preview.json',
  OpenExpoConfigPrebuild = 'expo.config.preview',
}

export enum PreviewModProvider {
  iosInfoPlist = 'ios.infoPlist',
  iosEntitlements = 'ios.entitlements',
  iosExpoPlist = 'ios.expoPlist',
  iosPodfileProperties = 'ios.podfileProperties',
  androidManifest = 'android.manifest',
  androidStrings = 'android.strings',
  androidColors = 'android.colors',
  androidColorsNight = 'android.colorsNight',
  androidStyles = 'android.styles',
  androidGradleProperties = 'android.gradleProperties',
}
