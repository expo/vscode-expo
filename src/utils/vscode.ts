import vscode, { CompletionItemProvider } from 'vscode';

import { ExpoProjectCache } from '../expo/project';

/**
 * Perform an asynchronous task with an "improvised" canellation token.
 * When cancelled, the task is either not started or the results aren't result.
 */
export async function withCancelToken<T>(
  token: vscode.CancellationToken,
  action: (token: vscode.CancellationToken) => Thenable<T>
): Promise<T | null> {
  if (token.isCancellationRequested) return null;
  const result = await action(token);
  return token.isCancellationRequested ? null : result;
}

export abstract class ExpoLinkProvider implements vscode.DocumentLinkProvider {
  constructor(
    { subscriptions }: vscode.ExtensionContext,
    protected projects: ExpoProjectCache,
    selector: vscode.DocumentSelector
  ) {
    subscriptions.push(vscode.languages.registerDocumentLinkProvider(selector, this));
  }

  abstract provideDocumentLinks(
    document: vscode.TextDocument,
    token: vscode.CancellationToken
  ): vscode.ProviderResult<vscode.DocumentLink[]>;
}

export abstract class ExpoDiagnosticsProvider {
  protected diagnostics: vscode.DiagnosticCollection;

  constructor(
    { subscriptions }: vscode.ExtensionContext,
    protected projects: ExpoProjectCache,
    private selector: vscode.DocumentSelector,
    diagnosticsName?: string
  ) {
    this.diagnostics = vscode.languages.createDiagnosticCollection(diagnosticsName);

    subscriptions.push(this.diagnostics);
    subscriptions.push(
      vscode.workspace.onDidSaveTextDocument((document) => this.diagnose(document))
    );
    subscriptions.push(
      vscode.window.onDidChangeActiveTextEditor((editor) => this.diagnose(editor?.document))
    );
  }

  public async diagnose(document?: vscode.TextDocument) {
    if (document && vscode.languages.match(this.selector, document)) {
      this.diagnostics.set(document.uri, await this.provideDiagnostics(document));
    } else if (document) {
      this.diagnostics.delete(document.uri);
    }
  }

  public abstract provideDiagnostics(
    document: vscode.TextDocument
  ): vscode.Diagnostic[] | Promise<vscode.Diagnostic[]>;
}

export abstract class ExpoCompletionsProvider implements CompletionItemProvider {
  constructor(
    { subscriptions }: vscode.ExtensionContext,
    protected projects: ExpoProjectCache,
    selector: vscode.DocumentSelector,
    triggerCharacters: string[]
  ) {
    subscriptions.push(
      vscode.languages.registerCompletionItemProvider(selector, this, ...triggerCharacters)
    );
  }

  abstract provideCompletionItems(
    document: vscode.TextDocument,
    position: vscode.Position,
    token: vscode.CancellationToken,
    context: vscode.CompletionContext
  ): vscode.ProviderResult<vscode.CompletionItem[] | vscode.CompletionList<vscode.CompletionItem>>;
}
