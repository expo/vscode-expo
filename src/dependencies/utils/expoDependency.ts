import { parse } from 'jsonc-parser';
import fetch from 'node-fetch';
import resolveFrom from 'resolve-from';
import semver from 'semver';
import { TextDecoder } from 'util';
import { Uri, workspace } from 'vscode';

export async function getExpoSdkVersion(projectRoot: string): Promise<string | undefined> {
  const expoPath = resolveFrom.silent(projectRoot, 'expo/package.json');
  if (!expoPath) {
    return undefined;
  }

  const expoBuffer = await workspace.fs.readFile(Uri.file(expoPath));
  const expoData = parse(new TextDecoder().decode(expoBuffer));
  if (!expoData) {
    return undefined;
  }

  return `${semver.major(expoData.version)}.0.0`;
}

export async function getVersionedDependencies(
  projectRoot: string,
  sdkVersion: string
): Promise<Record<string, string>> {
  const [localDependencies, remoteDependencies] = await Promise.all([
    fetchLocalVersionedDependencies(projectRoot).catch(() => ({})),
    fetchRemoteVersionedDependencies(sdkVersion).catch(() => ({})),
  ]);

  return {
    ...remoteDependencies,
    ...localDependencies,
  };
}

async function fetchRemoteVersionedDependencies(sdkVersion: string) {
  return fetch(`https://exp.host/--/api/v2/versions`)
    .then((response) => response.json())
    .then((data) => data.sdkVersions[sdkVersion].relatedPackages ?? {});
}

async function fetchLocalVersionedDependencies(projectRoot: string) {
  const dependenciesPath = resolveFrom.silent(projectRoot, 'expo/bundledNativeModules.json');
  if (!dependenciesPath) {
    return {};
  }

  const dependenciesBuffer = await workspace.fs.readFile(Uri.file(dependenciesPath));
  const dependenciesData = new TextDecoder().decode(dependenciesBuffer);

  return parse(dependenciesData) ?? {};
}
