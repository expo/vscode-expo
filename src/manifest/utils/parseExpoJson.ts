import { Node, ParseError, parseTree, findNodeAtLocation } from 'jsonc-parser';
import path from 'path';
import { TextDocument } from 'vscode';

let expoJsonCache: Record<string, Node> = {};

export const appJsonPattern = {
  scheme: 'file',
  // Match against app.json and app.config.json
  pattern: '**/*/app{,.config}.json',
  language: 'jsonc',
};

export function isAppJson(document: TextDocument) {
  return document && ['app.json', 'app.config.json'].includes(path.basename(document.fileName));
}

export function parseExpoJson(text: string): { node: Node | undefined; errors: ParseError[] } {
  if (text in expoJsonCache) {
    return { node: expoJsonCache[text], errors: [] };
  }

  const errors: ParseError[] = [];
  const tree = parseTree(text, errors);

  if (!tree) {
    return { errors, node: undefined };
  }

  const expo = findNodeAtLocation(tree, ['expo']);

  if (expo) {
    expoJsonCache = {
      [text]: expo,
    };
  }

  return { errors, node: expo };
}
