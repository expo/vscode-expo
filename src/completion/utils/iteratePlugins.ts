import { findNodeAtLocation, Node } from 'jsonc-parser';
import { Position, Range, TextDocument } from 'vscode';

import { parseExpoJson } from './parseExpoJson';

export interface JsonRange {
  offset: number;
  length: number;
}

export function findPluginsNode(appJson: Node | undefined) {
  return appJson ? findNodeAtLocation(appJson, ['plugins']) : null;
}

export function rangeForOffset(document: TextDocument, source: JsonRange) {
  return new Range(
    document.positionAt(source.offset),
    document.positionAt(source.offset + source.length)
  );
}

export function positionIsInPlugins(document: TextDocument, position: Position) {
  const { node } = parseExpoJson(document.getText());
  const pluginsNode = findPluginsNode(node);
  if (pluginsNode) {
    const range = rangeForOffset(document, pluginsNode);
    return range.contains(position);
  }

  return false;
}
