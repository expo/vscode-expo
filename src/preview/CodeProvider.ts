import { AndroidConfig, XML } from '@expo/config-plugins';
import plist from '@expo/plist';
import * as path from 'path';
import * as vscode from 'vscode';

import { clearProjectModules } from '../utils/clearModule';
import { getProjectRoot } from '../utils/getProjectRoot';

export type CodeProviderLanguage = 'json' | 'xml' | 'plist' | 'properties' | 'entitlements';

export type CodeProviderOptions = {
  convertLanguage?: CodeProviderLanguage;
  type: string;
};

const configurationSection = 'expo';

export type BasicCodeProviderOptions = Pick<CodeProviderOptions, 'convertLanguage'>;

export class CodeProvider implements vscode.TextDocumentContentProvider {
  readonly scheme: string = 'expo-config';

  public fileContents = '';
  public projectRoot: string;

  public readonly _onDidChange = new vscode.EventEmitter<vscode.Uri>();
  private readonly _changeSubscription: vscode.Disposable;
  private readonly _onDidChangeVisibleTextEditors: vscode.Disposable;
  private readonly _onDidChangeConfiguration: vscode.Disposable;

  public _isOpen = false;
  private _timer: NodeJS.Timer | undefined = undefined;
  private _editor: vscode.TextEditor | null = null;

  readonly defaultLanguage: CodeProviderLanguage = 'json';

  constructor(public _document: vscode.TextDocument, public options: CodeProviderOptions) {
    this.scheme = `expo-config-${this.options.type}`;
    this.projectRoot = getProjectRoot(this._document);

    this._changeSubscription = vscode.workspace.onDidChangeTextDocument((ev) => {
      this.textDidChange(ev);
      // Apply highlight to changed area
      if (this._editor && ev.document === this._editor.document) {
        const styles: vscode.DecorationOptions[] = [];
        for (const change of ev.contentChanges) {
          // Only style added lines
          if (!change.text || !change.range) {
            continue;
          }
          // Get the range of the modification
          const range = new vscode.Range(
            this._editor.document.positionAt(change.rangeOffset),
            this._editor.document.positionAt(change.rangeOffset + change.text.length)
          );
          styles.push({
            range,
          });
        }
        // Apply decorations (styles)
        this._editor.setDecorations(addedCodeStyle, styles);
      }
    });
    this._onDidChangeVisibleTextEditors = vscode.window.onDidChangeVisibleTextEditors((editors) =>
      this.visibleTextEditorsDidChange(editors)
    );
    this._onDidChangeConfiguration = vscode.workspace.onDidChangeConfiguration((ev) =>
      this.configurationDidChange(ev)
    );
  }

  setEditor(editor: vscode.TextEditor) {
    this._editor = editor;
  }

  getDefinedLanguage() {
    return this.options.convertLanguage || this.defaultLanguage;
  }

  /**
   * Where the preview file should be located.
   * The extension is used for code coloring, and having the file in the right location can enable the user to press cmd+s to save the file in place.
   */
  getFileName(): string {
    throw new Error('getFileName should be extended');
  }

  getURI() {
    const { convertLanguage } = this.options;
    let outputFileName = this.getFileName();
    const extension = path.extname(outputFileName).slice(1);
    if (extension !== this.defaultLanguage) {
      throw new Error('defaultLanguage should match fileName extension');
    }
    if (convertLanguage && convertLanguage !== this.defaultLanguage) {
      outputFileName += '.' + convertLanguage;
    }

    return vscode.Uri.parse(`${this.scheme}:${outputFileName}`);
  }

  formatWithLanguage(results: any, language?: CodeProviderLanguage) {
    language = language || this.options.convertLanguage || this.defaultLanguage;
    switch (language) {
      case 'json':
        return JSON.stringify(results, null, 2);
      case 'xml':
        return XML.format(results);
      case 'plist':
      case 'entitlements':
        return plist.build(results);
      case 'properties':
        return AndroidConfig.Properties.propertiesListToString(results);
      default:
        throw new Error('Unknown language: ' + language);
    }
  }

  sendDidChangeEvent() {
    if (!this._isOpen) return;
    this._onDidChange.fire(this.getURI());
  }

  dispose(): void {
    this._onDidChange.dispose();
    this._changeSubscription.dispose();
    this._onDidChangeVisibleTextEditors.dispose();
    this._onDidChangeConfiguration.dispose();
  }

  get document(): vscode.TextDocument {
    return this._document;
  }

  get documentName(): string {
    return this.getFileName();
  }

  setDocument(document: vscode.TextDocument): void {
    this._document = document;
  }

  get onDidChange(): vscode.Event<vscode.Uri> {
    return this._onDidChange.event;
  }

  private visibleTextEditorsDidChange(editors: Readonly<vscode.TextEditor[]>) {
    const isOpen = editors.some((e) => e.document.uri.scheme === this.scheme);
    if (!this._isOpen && isOpen) {
      this.update();
    }
    this._isOpen = isOpen;
  }

  private configurationDidChange(ev: vscode.ConfigurationChangeEvent): void {
    if (ev.affectsConfiguration(configurationSection)) {
      this.update();
    }
  }

  private textDidChange(ev: vscode.TextDocumentChangeEvent): void {
    if (!this._isOpen) return;

    if (ev.document !== this._document) return;

    if (this._timer) {
      clearTimeout(this._timer);
    }
    this._timer = setTimeout(() => {
      this._timer = undefined;
      this.update();
    }, 300);
  }

  async update(): Promise<void> {
    try {
      // Reset all requires to ensure plugins update
      clearProjectModules(this.projectRoot);

      this.fileContents = this.formatWithLanguage(await this.getFileContents());
    } catch (error) {
      this.fileContents = '';
      vscode.window.showErrorMessage(error.message);
    }

    this.sendDidChangeEvent();
  }

  getExpoConfigAsync() {}

  provideTextDocumentContent(
    _uri: vscode.Uri,
    _token: vscode.CancellationToken
  ): vscode.ProviderResult<string> {
    this._isOpen = true;

    return this.fileContents;
  }

  async getFileContents(): Promise<any> {
    return '';
  }
}

// create a decorator type that we use to decorate small numbers
const addedCodeStyle = vscode.window.createTextEditorDecorationType({
  //   before: {
  //     contentText: '+',
  //   },
  isWholeLine: true,
  backgroundColor: '#e5ffec',
  dark: {
    backgroundColor: '#2ea04333',
  },
});
