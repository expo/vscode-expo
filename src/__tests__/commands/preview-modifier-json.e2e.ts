import { expect } from 'chai';
import { commands, window } from 'vscode';

import { PreviewCommand, PreviewModProvider } from '../../preview/constants';
import { sanitizeSnapshotValues } from '../utils/snapshot';
import { closeAllEditors, getWorkspaceUri, waitForEditorOpen } from '../utils/vscode';

describe(PreviewCommand.OpenExpoFileJsonPrebuild, () => {
  beforeEach(async () => {
    await window.showTextDocument(getWorkspaceUri('preview/app.json'));
  });

  afterEach(async () => {
    await closeAllEditors();
  });

  it(`runs for ${PreviewModProvider.androidColors} in json format`, async () => {
    await commands.executeCommand(
      PreviewCommand.OpenExpoFileJsonPrebuild,
      PreviewModProvider.androidColors
    );

    const preview = await waitForEditorOpen('colors.xml.json');
    expect(preview).to.exist;

    const content = sanitizeSnapshotValues(preview!.document.getText());
    expect(content).toMatchSnapshot();
  });

  it(`runs for ${PreviewModProvider.androidColorsNight} in json format`, async () => {
    await commands.executeCommand(
      PreviewCommand.OpenExpoFileJsonPrebuild,
      PreviewModProvider.androidColorsNight
    );

    const preview = await waitForEditorOpen('colors.xml.json');
    expect(preview).to.exist;

    const content = sanitizeSnapshotValues(preview!.document.getText());
    expect(content).toMatchSnapshot();
  });

  it(`runs for ${PreviewModProvider.androidGradleProperties} in json format`, async () => {
    await commands.executeCommand(
      PreviewCommand.OpenExpoFileJsonPrebuild,
      PreviewModProvider.androidGradleProperties
    );

    const preview = await waitForEditorOpen('gradle.properties.json');
    expect(preview).to.exist;

    const content = sanitizeSnapshotValues(preview!.document.getText());
    expect(content).toMatchSnapshot();
  });

  it(`runs for ${PreviewModProvider.androidManifest} in json format`, async () => {
    await commands.executeCommand(
      PreviewCommand.OpenExpoFileJsonPrebuild,
      PreviewModProvider.androidManifest
    );

    const preview = await waitForEditorOpen('AndroidManifest.xml.json');
    expect(preview).to.exist;

    const content = sanitizeSnapshotValues(preview!.document.getText());
    expect(content).toMatchSnapshot();
  });

  it(`runs for ${PreviewModProvider.androidStrings} in json format`, async () => {
    await commands.executeCommand(
      PreviewCommand.OpenExpoFileJsonPrebuild,
      PreviewModProvider.androidStrings
    );

    const preview = await waitForEditorOpen('strings.xml.json');
    expect(preview).to.exist;

    const content = sanitizeSnapshotValues(preview!.document.getText());
    expect(content).toMatchSnapshot();
  });

  it(`runs for ${PreviewModProvider.androidStyles} in json format`, async () => {
    await commands.executeCommand(
      PreviewCommand.OpenExpoFileJsonPrebuild,
      PreviewModProvider.androidStyles
    );

    const preview = await waitForEditorOpen('styles.xml.json');
    expect(preview).to.exist;

    const content = sanitizeSnapshotValues(preview!.document.getText());
    expect(content).toMatchSnapshot();
  });

  it(`runs for ${PreviewModProvider.iosEntitlements} in json format`, async () => {
    await commands.executeCommand(
      PreviewCommand.OpenExpoFileJsonPrebuild,
      PreviewModProvider.iosEntitlements
    );

    const preview = await waitForEditorOpen('Example.entitlements.json');
    expect(preview).to.exist;

    const content = sanitizeSnapshotValues(preview!.document.getText());
    expect(content).toMatchSnapshot();
  });

  it(`runs for ${PreviewModProvider.iosExpoPlist} in json format`, async () => {
    await commands.executeCommand(
      PreviewCommand.OpenExpoFileJsonPrebuild,
      PreviewModProvider.iosExpoPlist
    );

    const preview = await waitForEditorOpen('Expo.plist.json');
    expect(preview).to.exist;

    const content = sanitizeSnapshotValues(preview!.document.getText());
    expect(content).toMatchSnapshot();
  });

  it(`runs for ${PreviewModProvider.iosInfoPlist} in json format`, async () => {
    await commands.executeCommand(
      PreviewCommand.OpenExpoFileJsonPrebuild,
      PreviewModProvider.iosInfoPlist
    );

    const preview = await waitForEditorOpen('Info.plist.json');
    expect(preview).to.exist;

    const content = sanitizeSnapshotValues(preview!.document.getText());
    expect(content).toMatchSnapshot();
  });

  it(`runs for ${PreviewModProvider.iosPodfileProperties} in json format`, async () => {
    await commands.executeCommand(
      PreviewCommand.OpenExpoFileJsonPrebuild,
      PreviewModProvider.iosPodfileProperties
    );

    const preview = await waitForEditorOpen('Podfile.properties.json');
    expect(preview).to.exist;

    const content = sanitizeSnapshotValues(preview!.document.getText());
    expect(content).toMatchSnapshot();
  });
});
