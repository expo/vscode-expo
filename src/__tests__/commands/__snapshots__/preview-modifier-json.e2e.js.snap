// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`expo.config.prebuild.preview.json runs for android.colors in json format 1`] = `
"{
  \\"resources\\": {
    \\"color\\": [
      {
        \\"$\\": {
          \\"name\\": \\"iconBackground\\"
        },
        \\"_\\": \\"#FFFFFF\\"
      },
      {
        \\"$\\": {
          \\"name\\": \\"colorPrimary\\"
        },
        \\"_\\": \\"#023c69\\"
      },
      {
        \\"$\\": {
          \\"name\\": \\"colorPrimaryDark\\"
        },
        \\"_\\": \\"#ffffff\\"
      },
      {
        \\"$\\": {
          \\"name\\": \\"splashscreen_background\\"
        },
        \\"_\\": \\"#ffffff\\"
      }
    ]
  }
}"
`;

exports[`expo.config.prebuild.preview.json runs for android.colorsNight in json format 1`] = `
"{
  \\"resources\\": {}
}"
`;

exports[`expo.config.prebuild.preview.json runs for android.gradleProperties in json format 1`] = `
"{
  \\"0\\": {
    \\"type\\": \\"property\\",
    \\"key\\": \\"expo.jsEngine\\",
    \\"value\\": \\"jsc\\"
  }
}"
`;

exports[`expo.config.prebuild.preview.json runs for android.manifest in json format 1`] = `
"{
  \\"manifest\\": {
    \\"uses-permission\\": [
      {
        \\"$\\": {
          \\"android:name\\": \\"android.permission.INTERNET\\"
        }
      },
      {
        \\"$\\": {
          \\"android:name\\": \\"android.permission.SYSTEM_ALERT_WINDOW\\"
        }
      },
      {
        \\"$\\": {
          \\"android:name\\": \\"android.permission.READ_EXTERNAL_STORAGE\\"
        }
      },
      {
        \\"$\\": {
          \\"android:name\\": \\"android.permission.WRITE_EXTERNAL_STORAGE\\"
        }
      }
    ],
    \\"queries\\": [
      {
        \\"intent\\": [
          {
            \\"action\\": [
              {
                \\"$\\": {
                  \\"android:name\\": \\"android.intent.action.VIEW\\"
                }
              }
            ],
            \\"category\\": [
              {
                \\"$\\": {
                  \\"android:name\\": \\"android.intent.category.BROWSABLE\\"
                }
              }
            ],
            \\"data\\": [
              {
                \\"$\\": {
                  \\"android:scheme\\": \\"https\\"
                }
              }
            ]
          }
        ]
      }
    ],
    \\"application\\": [
      {
        \\"meta-data\\": [
          {
            \\"$\\": {
              \\"android:name\\": \\"expo.modules.updates.EXPO_UPDATE_URL\\",
              \\"android:value\\": \\"https://exp.host/@bycedric/expo-app\\"
            }
          },
          {
            \\"$\\": {
              \\"android:name\\": \\"expo.modules.updates.EXPO_SDK_VERSION\\",
              \\"android:value\\": \\"45.0.0\\"
            }
          },
          {
            \\"$\\": {
              \\"android:name\\": \\"expo.modules.updates.ENABLED\\",
              \\"android:value\\": \\"true\\"
            }
          },
          {
            \\"$\\": {
              \\"android:name\\": \\"expo.modules.updates.EXPO_UPDATES_CHECK_ON_LAUNCH\\",
              \\"android:value\\": \\"ALWAYS\\"
            }
          },
          {
            \\"$\\": {
              \\"android:name\\": \\"expo.modules.updates.EXPO_UPDATES_LAUNCH_WAIT_MS\\",
              \\"android:value\\": \\"0\\"
            }
          }
        ],
        \\"activity\\": [
          {
            \\"$\\": {
              \\"android:name\\": \\".MainActivity\\",
              \\"android:label\\": \\"@string/app_name\\",
              \\"android:configChanges\\": \\"keyboard|keyboardHidden|orientation|screenSize|uiMode\\",
              \\"android:launchMode\\": \\"singleTask\\",
              \\"android:windowSoftInputMode\\": \\"adjustResize\\",
              \\"android:theme\\": \\"@style/Theme.App.SplashScreen\\",
              \\"android:screenOrientation\\": \\"portrait\\"
            },
            \\"intent-filter\\": [
              {
                \\"action\\": [
                  {
                    \\"$\\": {
                      \\"android:name\\": \\"android.intent.action.MAIN\\"
                    }
                  }
                ],
                \\"category\\": [
                  {
                    \\"$\\": {
                      \\"android:name\\": \\"android.intent.category.LAUNCHER\\"
                    }
                  }
                ]
              },
              {
                \\"action\\": [
                  {
                    \\"$\\": {
                      \\"android:name\\": \\"android.intent.action.VIEW\\"
                    }
                  }
                ],
                \\"category\\": [
                  {
                    \\"$\\": {
                      \\"android:name\\": \\"android.intent.category.DEFAULT\\"
                    }
                  },
                  {
                    \\"$\\": {
                      \\"android:name\\": \\"android.intent.category.BROWSABLE\\"
                    }
                  }
                ],
                \\"data\\": [
                  {
                    \\"$\\": {
                      \\"android:scheme\\": \\"com.placeholder.appid\\"
                    }
                  }
                ]
              }
            ]
          },
          {
            \\"$\\": {
              \\"android:name\\": \\"com.facebook.react.devsupport.DevSettingsActivity\\"
            }
          }
        ],
        \\"$\\": {
          \\"android:name\\": \\".MainApplication\\",
          \\"android:label\\": \\"@string/app_name\\",
          \\"android:icon\\": \\"@mipmap/ic_launcher\\",
          \\"android:roundIcon\\": \\"@mipmap/ic_launcher_round\\",
          \\"android:allowBackup\\": \\"true\\",
          \\"android:theme\\": \\"@style/AppTheme\\",
          \\"android:usesCleartextTraffic\\": \\"true\\"
        }
      }
    ],
    \\"$\\": {
      \\"xmlns:android\\": \\"http://schemas.android.com/apk/res/android\\",
      \\"package\\": \\"com.placeholder.appid\\"
    }
  }
}"
`;

exports[`expo.config.prebuild.preview.json runs for android.strings in json format 1`] = `
"{
  \\"resources\\": {
    \\"string\\": [
      {
        \\"$\\": {
          \\"name\\": \\"app_name\\"
        },
        \\"_\\": \\"expo-app\\"
      },
      {
        \\"$\\": {
          \\"name\\": \\"expo_splash_screen_resize_mode\\",
          \\"translatable\\": \\"false\\"
        },
        \\"_\\": \\"contain\\"
      },
      {
        \\"$\\": {
          \\"name\\": \\"expo_splash_screen_status_bar_translucent\\",
          \\"translatable\\": \\"false\\"
        },
        \\"_\\": \\"false\\"
      },
      {
        \\"$\\": {
          \\"name\\": \\"expo_system_ui_user_interface_style\\",
          \\"translatable\\": \\"false\\"
        },
        \\"_\\": \\"light\\"
      }
    ]
  }
}"
`;

exports[`expo.config.prebuild.preview.json runs for android.styles in json format 1`] = `
"{
  \\"resources\\": {
    \\"$\\": {
      \\"xmlns:tools\\": \\"http://schemas.android.com/tools\\"
    },
    \\"style\\": [
      {
        \\"$\\": {
          \\"name\\": \\"AppTheme\\",
          \\"parent\\": \\"Theme.AppCompat.Light.NoActionBar\\"
        },
        \\"item\\": [
          {
            \\"$\\": {
              \\"name\\": \\"colorPrimary\\"
            },
            \\"_\\": \\"@color/colorPrimary\\"
          },
          {
            \\"$\\": {
              \\"name\\": \\"colorPrimaryDark\\"
            },
            \\"_\\": \\"@color/colorPrimaryDark\\"
          }
        ]
      },
      {
        \\"$\\": {
          \\"name\\": \\"Theme.App.SplashScreen\\",
          \\"parent\\": \\"AppTheme\\"
        },
        \\"item\\": [
          {
            \\"$\\": {
              \\"name\\": \\"android:windowBackground\\"
            },
            \\"_\\": \\"@drawable/splashscreen\\"
          }
        ]
      }
    ]
  }
}"
`;

exports[`expo.config.prebuild.preview.json runs for ios.entitlements in json format 1`] = `
"{
  \\"aps-environment\\": \\"development\\"
}"
`;

exports[`expo.config.prebuild.preview.json runs for ios.expoPlist in json format 1`] = `
"{
  \\"EXUpdatesCheckOnLaunch\\": \\"ALWAYS\\",
  \\"EXUpdatesEnabled\\": true,
  \\"EXUpdatesLaunchWaitMs\\": 0,
  \\"EXUpdatesSDKVersion\\": \\"45.0.0\\",
  \\"EXUpdatesURL\\": \\"https://exp.host/@bycedric/expo-app\\"
}"
`;

exports[`expo.config.prebuild.preview.json runs for ios.infoPlist in json format 1`] = `
"{
  \\"CFBundleDevelopmentRegion\\": \\"$(DEVELOPMENT_LANGUAGE)\\",
  \\"CFBundleDisplayName\\": \\"expo-app\\",
  \\"CFBundleExecutable\\": \\"$(EXECUTABLE_NAME)\\",
  \\"CFBundleIdentifier\\": \\"$(PRODUCT_BUNDLE_IDENTIFIER)\\",
  \\"CFBundleInfoDictionaryVersion\\": \\"6.0\\",
  \\"CFBundleName\\": \\"$(PRODUCT_NAME)\\",
  \\"CFBundlePackageType\\": \\"$(PRODUCT_BUNDLE_PACKAGE_TYPE)\\",
  \\"CFBundleShortVersionString\\": \\"1.0.0\\",
  \\"CFBundleSignature\\": \\"????\\",
  \\"CFBundleURLTypes\\": [
    {
      \\"CFBundleURLSchemes\\": [
        \\"com.placeholder.appid\\"
      ]
    }
  ],
  \\"CFBundleVersion\\": \\"1\\",
  \\"LSRequiresIPhoneOS\\": true,
  \\"NSAppTransportSecurity\\": {
    \\"NSAllowsArbitraryLoads\\": true,
    \\"NSExceptionDomains\\": {
      \\"localhost\\": {
        \\"NSExceptionAllowsInsecureHTTPLoads\\": true
      }
    }
  },
  \\"UILaunchStoryboardName\\": \\"SplashScreen\\",
  \\"UIRequiredDeviceCapabilities\\": [
    \\"armv7\\"
  ],
  \\"UIRequiresFullScreen\\": false,
  \\"UIStatusBarStyle\\": \\"UIStatusBarStyleDefault\\",
  \\"UISupportedInterfaceOrientations\\": [
    \\"UIInterfaceOrientationPortrait\\",
    \\"UIInterfaceOrientationPortraitUpsideDown\\"
  ],
  \\"UISupportedInterfaceOrientations~ipad\\": [
    \\"UIInterfaceOrientationPortrait\\",
    \\"UIInterfaceOrientationPortraitUpsideDown\\",
    \\"UIInterfaceOrientationLandscapeLeft\\",
    \\"UIInterfaceOrientationLandscapeRight\\"
  ],
  \\"UIUserInterfaceStyle\\": \\"Light\\",
  \\"UIViewControllerBasedStatusBarAppearance\\": false
}"
`;

exports[`expo.config.prebuild.preview.json runs for ios.podfileProperties in json format 1`] = `
"{
  \\"expo.jsEngine\\": \\"jsc\\"
}"
`;
